### PR TITLE
[FLINK-22698][connectors/rabbitmq] Add deliveryTimeout parameter to RabbitMQ source

### PR DESCRIPTION
**Pull request by cmick**
## What is the purpose of the change

This pull request enables setting the next message delivery timeout in the RabbitMQ queueing consumer, allowing to properly stop the job in cases when no new message is available on the queue (issue [FLINK-22698](https://issues.apache.org/jira/browse/FLINK-22698)).

Currently, the issue described FLINK-22698 causes the job to never stop (gracefully via stop command) if there is no new data. Since the current behavior (no timeout) is causing trouble, this PR introduces setting a default value of 30s. The upper bound is dictated by the default Flink `client.timeout` (60s) - RMQ connector should not be the cause of CLI timeouts. The lower bound is related to CPU utilization - if set to some really low values (such as 1ms), it could start to have a noticeable effect (redundant queue check every 1ms).

## Brief change log

 - adds the ability to `setDeliveryTimeout` on the `RMQConnectionConfig` and its builder
 - change default message delivery timeout in the `RMQSource` to 30 seconds (previously there was no timeout)

## Verifying this change

This change added tests and can be verified as follows:
 - `RMQSourceTest#testDeliveryTimeout`, which tests correct cancelling of the source when no new messages are available
 - `RMQConnectionConfigTest#shouldSetDeliveryTimeout`, which tests setting a custom delivery timeout
 - `RMQConnectionConfigTest#shouldReturnDefaultDeliveryTimeout`, which tests using a default delivery timeout, when custom is not provided
 - `RMQConnectionConfigTest#shouldThrowIllegalArgumentExceptionIfDeliveryTimeoutIsNegative` and `RMQConnectionConfigTest#shouldThrowIllegalArgumentExceptionIfDeliveryTimeoutWithUnitIsNegative`, which tests against invalid timeout values

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive):  **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper:  **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **JavaDocs**

### CODE REVIEW

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java
Comment (SteNicholas) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java:333]: Should this default value for delivery timeout put into the constructor of `RMQConnectionConfig` ?

**Comment by SteNicholas** [comment]
@cmick , IMO, you should rerun the CI for your pull request. Current CI result is failed and hasn't run the RabbitMQ connector tests.

**Comment by cmick** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java]: Thanks. I've moved the default value setting to `RMQConnectionConfig` (14571a9)

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (SteNicholas) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:104]: if deliveryTimeout is null, the deliveryTimeout could be `DEFAULT_DELIVERY_TIMEOUT`. Right?

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (SteNicholas) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:544]: Why doesn't this check whether deliveryTimeout is negative?

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (SteNicholas) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: Why is the default value of delivery timeout is 30 seconds?

Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: We decided in the discussion under the related ticket (https://issues.apache.org/jira/browse/FLINK-22698) to set a sensible default that will resolve the potential problem that a user may encounter while stopping the job (the other option is to just leave the current problematic behavior by setting in to 0).
There's nothing special about 30s, but I would say it should be lower than 60s as this is the default timeout for the CLI client stop command.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:546]: It's already checked in the constructor (invoked in the build method), but it can be checked here as well to indicate invalid value where it's actually set. Added input validation in f90e2d2 - thanks

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (SteNicholas) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: My point of concern is why the default value is 30 seconds, and what is the basis for the default value of 30 seconds.

Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: Yes, I understand. The issue described FLINK-22698 causes the job to never stop (gracefully via stop command) if auto-ack is disabled and there is no new data. So, the current behavior (no timeout) is causing trouble (that's why its better to set some default here).
IMO the upper bound is dictated by the default Flink _client.timeout_ (60s) - RMQ connector should not be the cause of CLI timeouts. The lower bound is related to CPU utilization - if set to some really low values (such as 1ms), it could start to have a noticeable effect (redundant queue check every 1ms).
Again, there is nothing special with the 30s (e.g. we can set it to 15s as well), but it **fixes the issue** and meets the above criteria.

Comment (SteNicholas) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: Thanks for your detailed explanation. I got your point for the delivery timeout and makes sense to me.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (SteNicholas) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:370]: IMO, the type of `deliveryTimeout` is `Long`.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java]: Great, thanks. I updated the PR description with the details above if anyone else was looking for a rationale

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java]: Agree, it sounds more intuitive, thanks. Changed in 54b08d3

**Review by SteNicholas — APPROVED** [review]

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (austince) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:106]: Can we split this into two checks or update the error message to account for null?

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (austince) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:149]: Same comment here about splitting the checks or updating the message.

**Review by austince — CHANGES_REQUESTED** [review]
Thanks for the contribution + thanks for the review work @SteNicholas! Looks good to me, bar a small nit. I'll look for a committer who can do the final ok and merge this as well in the meantime.

**Review by austince — APPROVED** [review]

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:106]: I think this check should be 
```suggestion
        Preconditions.checkArgument(
                deliveryTimeout != null && deliveryTimeout >= 0,
                "deliveryTimeout can not be negative");
```

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:149]: Same comment as above.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: I'd prefer to have this value as part of the `RMQSource` class. This way this class does not hold any logic and it can be easier reused.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:546]: I agree with @austince's comment.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:538]: Nit: If you move the constant variable out of this class please also update the comments.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:553]: Nit: update comment if constant is moved.

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:424]: Nit: Can you move this helper after all `@Test` methods?

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:450]: In general, I'd recommend not relying on time in test cases because it may take longer on other systems to start the source which may introduce a test instability.

**Review by fapaul — COMMENTED** [review]
Thanks for adding the `deliveryTimeout` feature. Perhaps I am not fully understanding the problem yet because in the JIRA you are describing that canceling the job still works and **only** stop with savepoint deadlocks.

If I am not missing something it should also not be possible to cancel the job until something was delivered which would mean the current solution does not fully suffice. 
@cmick Can you explain this?

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:48]: @SteNicholas What was your reasoning to put it here?

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:106]: The check is correct. The value should either be not set at all (so `null` is fine, default will be used here) or set to non-negative number.
It could be refactored a bit, so that the config would `long` instead of nullable `Long` and the default (if not provided) would be set by the builder. But I would have doubts if this is a good idea. Currently the builder is just a helper to build the config instance, with no internal logic.
What's your opinion? Or maybe you have some other ideas how to make it more clear?

**Review by cmick — COMMENTED** [review]
> ... Perhaps I am not fully understanding the problem yet because in the JIRA you are describing that canceling the job still works and **only** stop with savepoint deadlocks.
> 
> If I am not missing something it should also not be possible to cancel the job until something was delivered which would mean the current solution does not fully suffice.

@fapaul Savepoint is completed correctly (so it's not a savepoint deadlock). But the job does not terminate after completing the checkpoint (it should). The difference is how `stop` and `cancel` commands are handled by Flink. Cancelling the job works because it **interrupts** the operator threads (so the whole consumer thread that is waiting for the next message is interrupted). Stop command tries to terminate the job **gracefully** (waiting for the operators to finish after invoking the `cancel` method).

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:450]: You're right. This would require extending the test into:
1) generating and consuming at least 1 message (we need to make sure that the consumer was started), instead of waiting arbitrary 5ms 
2) stopping the message delivery (similarly as its currently implemented, but after the source was started)
3) cancelling the source and waiting for it to stop (as its currently implemented)
4) checking if the job stopped gracefully (this can be verified by adding some helper boolean flag) instead of checking the exception (so this would eliminate the second Thread.sleep).

Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java]: I refactored the delivery timeout unit test so that it does not rely on any timeouts or processing time at all (changes in a229e25). I also moved the helper class after all `@Test` methods.

**Comment by cmick** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java:331]: Can the `timeout` actually be below zero? In general, I would not allow calling `consumer.nextDelivery()` at all because it prevents the job from making progress.

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:431]: I think with the current setup we have for the RabbitMQ source it is very tough to test the added functionality. We would need to access the underlying `queue` in the Consumer to mock the behavior. 
My suggestion would be to change this test slightly to ensure `consumer.nextDelivery()` is never called.

**Review by fapaul — COMMENTED** [review]
When looking at the testing setup again it is, unfortunately, a bit limited and the mocks are not much help in your case. I proposed a solution let me know what you think.

Overall, it looks almost ready.

**Comment by fapaul** [comment]
@cmick I have set up a proper integration and added a test for the stop with savepoint capability https://github.com/apache/flink/commit/661d6f8bd17fdf1d5c0307aea1161f374006151e you can just cherry-pick that commit and replace your current test. Hopefully, this simplifies your testing effort. 

I think this feature/bug is very critical and we should backport it to release-1.13/release-1.12, too.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java]: Changed in 0b2a495

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java:331]: It can be set to zero. Zero is the current, problematic, behavior (no timeout). Any positive number will enable the timeout (and it will be enabled by default). Do you mean, the "no timeout" case should be dropped completely?

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:427]: Thanks. I've cherry-picked the integration tests you implemented. Looks great (I've just removed a duplicated dependency - 12ea915).
I've changed (b86d622) the deliveryTimeout test (it now basically just checks whether the correct method was called).

**Comment by cmick** [comment]
@fapaul I think I covered all the things discussed, but let me know if anything else should be improved.

Backport to 1.13 and 1.12 should be straightforward as the connector did not change much (if at all). I can help with this as well

**Comment by cmick** [comment]
@flinkbot run azure

**Comment by cmick** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java:331]: IMO we should change the validation slightly to always require a value greater than 0. I assume that most RabbitMQ uses cases are not very performance-critical and can easily take a small penalty.

Preventing the bug from happening is the highest priority.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:105]: As a followup to not allow shooting themself in the foot.
```suggestion
                deliveryTimeout == null || deliveryTimeout > 0,
```

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:149]:
```suggestion
                deliveryTimeout == null || deliveryTimeout > 0,
```

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (fapaul) [inline flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java:547]:
```suggestion
                    deliveryTimeout > 0, "deliveryTimeout can not be negative");
```

File: flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java
Comment (fapaul) [reply flink-connectors/flink-connector-rabbitmq/src/test/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSourceTest.java:427]: Thanks, and I am a bit sorry for the missed styling check but you've already fixed it 👍

**Review by fapaul — COMMENTED** [review]
As last thing: the delivery timeout should not be allowed to be 0 to fully prevent the job from being stuck.

File: flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java
Comment (cmick) [reply flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/common/RMQConnectionConfig.java]: You're right. I've applied the change everywhere applicable (config, source and docs) in fa7bcb9. Now only the `consumer.nextDelivery(timeout)` is called and it's not possible to set the timeout to zero.

**Review by fapaul — APPROVED** [review]
Thanks for your contribution. Can you squash all the commits and then open the backports to the release-1.13 and release-1.12 branch?

We will merge the PRs once the CI is green.

**Comment by cmick** [comment]
@fapaul thanks for your review and help. I've squashed all commits and rebased on current master.

I've created a backport for 1.13 (#16222), although I have some troubles with the integration test on 1.12 (the `getRestClusterClient` method  does not exist). Can we just skip the IT for 1.12 or should I try to refactor it?

**Comment by cmick** [comment]
@flinkbot run azure

**Comment by fapaul** [comment]
@cmick I think it is okay to skip the IT case for 1.12. I briefly checked and I do not think it is worth the effort to change the test significantly.

**Comment by cmick** [comment]
Actually I managed to do it without much of an effort (a single line had to be changed, as one of the helper methods did not exist in 1.12). BP for 1.12: #16223

**Comment by fapaul** [comment]
Perfect, thanks for looking into it.

**Comment by cmick** [comment]
@flinkbot run azure

**Comment by fapaul** [comment]
Test failures seem unrelated https://issues.apache.org/jira/projects/FLINK/issues/FLINK-22890?filter=allopenissues

**Comment by fapaul** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
Thank you very much for your contribution. Failure unrelated, so I'm merging this and also the backports.

**Pull request merged**
